main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintGen.scala

```diff
@@ -478,14 +478,19 @@ object ConstraintGen {
 
       case Expr.ExtMatch(exp, rules, loc) =>
         val (tpe, eff) = visitExp(exp)
-        val (caseTpes, tpes, effs) = rules.map(visitExtMatchRule).unzip3
-        val expectedRowType = caseTpes.foldRight(Type.mkSchemaRowEmpty(loc)) {
-          case ((name, patTpes), acc) => Type.mkSchemaRowExtend(name, Type.mkRelation(patTpes, loc), acc, loc)
-        }
-        val expectedSchemaType = Type.mkExtensible(expectedRowType, loc)
-        c.unifyType(tpe, expectedSchemaType, exp.loc)
-        c.unifyAllTypes(tpes, loc)
-        val resTpe = tpes.head // Note: We are guaranteed to have one rule.
+        val (patTypes, branchExpTypes, effs) = rules.map(visitExtMatchRule).unzip3
+        val expectedRowType =
+          patTypes.collect { case Left(value) => value }
+            .foldRight(Type.mkSchemaRowEmpty(loc.asSynthetic)) {
+              case ((pred, tpes), acc) =>
+                val relation = Type.mkRelation(tpes, pred.loc.asSynthetic)
+                Type.mkSchemaRowExtend(pred, relation, acc, pred.loc.asSynthetic)
+            }
+        val expectedExtensibleType = Type.mkExtensible(expectedRowType, loc.asSynthetic)
+        val tvars = patTypes.collect { case Right(value) => value }
```

Comment:
I think some better naming in the above code. For example, here tvars is really defaultTypes or similar.